File: plugins/cookbook/sidecar/python/hello-world/test_hello.py
#!/usr/bin/env python3
"""
Subprocess test for hello.py — exercises the stdin/stdout JSON-RPC
contract without the Smiths-Net engine.

Run:
    python3 test_hello.py
"""

from __future__ import annotations

import json
import subprocess
import sys
from pathlib import Path

SCRIPT = Path(__file__).parent / "hello.py"


def rpc(proc: subprocess.Popen, req_id: int, method: str, params: dict | None = None) -> dict:
    """Send one JSON-RPC request and read the response."""
    request = {
        "jsonrpc": "2.0",
        "id": req_id,
        "method": method,
        "params": params or {},
    }
    frame = json.dumps(request, separators=(",", ":")) + "\n"
    proc.stdin.write(frame)
    proc.stdin.flush()

    line = proc.stdout.readline()
    assert line, f"expected response for {method}, got EOF"
    return json.loads(line)


def test_describe_capabilities() -> None:
    proc = subprocess.Popen(
        [sys.executable, str(SCRIPT)],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )

    try:
        resp = rpc(proc, 1, "describe_capabilities")
        assert resp["id"] == 1, f"id mismatch: {resp}"
        assert "result" in resp, f"missing result: {resp}"
        result = resp["result"]
        assert result["name"] == "hello-world-py", f"bad name: {result}"
        assert result["version"] == "0.1.0", f"bad version: {result}"
        print("✓ describe_capabilities")
    finally:
        proc.terminate()
        proc.wait()


def test_echo_unknown_method() -> None:
    proc = subprocess.Popen(
        [sys.executable, str(SCRIPT)],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )

    try:
        resp = rpc(proc, 2, "some_custom_method")
        assert resp["id"] == 2, f"id mismatch: {resp}"
        assert resp["result"]["echo"] == "some_custom_method", f"bad echo: {resp}"
        print("✓ echo unknown method")
    finally:
        proc.terminate()
        proc.wait()


def test_shutdown() -> None:
    proc = subprocess.Popen(
        [sys.executable, str(SCRIPT)],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )

    resp = rpc(proc, 3, "shutdown")
    assert resp["id"] == 3, f"id mismatch: {resp}"
    assert resp["result"]["ok"] is True, f"bad shutdown result: {resp}"

    # Process should exit cleanly after shutdown.
    rc = proc.wait(timeout=5)
    assert rc == 0, f"expected exit 0, got {rc}"
    print("✓ shutdown")


def test_multiple_requests() -> None:
    proc = subprocess.Popen(
        [sys.executable, str(SCRIPT)],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )

    try:
        r1 = rpc(proc, 10, "describe_capabilities")
        assert r1["id"] == 10
        r2 = rpc(proc, 11, "ping")
        assert r2["id"] == 11
        assert r2["result"]["echo"] == "ping"
        r3 = rpc(proc, 12, "another")
        assert r3["id"] == 12
        assert r3["result"]["echo"] == "another"
        print("✓ multiple sequential requests")
    finally:
        proc.terminate()
        proc.wait()


if __name__ == "__main__":
    test_describe_capabilities()
    test_echo_unknown_method()
    test_shutdown()
    test_multiple_requests()
    print("\nAll tests passed.")

File: plugins/cookbook/sidecar/nodejs/hello-world/test_hello.mjs
#!/usr/bin/env node
/**
 * Subprocess test for hello.mjs — exercises the stdin/stdout JSON-RPC
 * contract without the Smiths-Net engine.
 */

import { spawn } from "node:child_process";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PLUGIN = join(__dirname, "hello.mjs");

let passed = 0;
let failed = 0;

/**
 * Send a JSON-RPC request and read the response.
 * @param {object} request
 * @returns {Promise<object>}
 */
function rpc(proc, request) {
  return new Promise((resolve, reject) => {
    let buffer = "";

    const onData = (chunk) => {
      buffer += chunk.toString();
      const lines = buffer.split("\n");
      for (const line of lines) {
        const trimmed = line.trim();
        if (!trimmed) continue;
        try {
          const response = JSON.parse(trimmed);
          proc.stdout.removeListener("data", onData);
          resolve(response);
          return;
        } catch {
          // partial line, keep buffering
        }
      }
    };

    proc.stdout.on("data", onData);
    proc.stdin.write(JSON.stringify(request) + "\n");

    setTimeout(() => reject(new Error("timeout")), 3000);
  });
}

function assert(condition, label) {
  if (condition) {
    console.log(`✓ ${label}`);
    passed++;
  } else {
    console.error(`✗ ${label}`);
    failed++;
  }
}

async function main() {
  const proc = spawn("node", [PLUGIN], {
    stdio: ["pipe", "pipe", "pipe"],
  });

  // Test 1: describe_capabilities
  const desc = await rpc(proc, {
    jsonrpc: "2.0",
    id: 1,
    method: "describe_capabilities",
    params: {},
  });
  assert(desc.result?.name === "hello-world-nodejs", "describe_capabilities");

  // Test 2: echo unknown method
  const echo = await rpc(proc, {
    jsonrpc: "2.0",
    id: 2,
    method: "ping",
    params: {},
  });
  assert(echo.result?.echo === "ping", "echo unknown method");

  // Test 3: shutdown
  const shut = await rpc(proc, {
    jsonrpc: "2.0",
    id: 3,
    method: "shutdown",
    params: {},
  });
  assert(shut.result?.ok === true, "shutdown");

  // Wait for process to exit.
  await new Promise((resolve) => proc.on("close", resolve));

  console.log(`\n${passed} passed, ${failed} failed`);
  process.exit(failed > 0 ? 1 : 0);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
